src/scripting/lua_menu_item.cpp

```diff
@@ -53,14 +56,22 @@ lua_ptr<gui2::menu_item>& luaW_checkmenuitem_ptr(lua_State* L, int n)
 
 static int impl_menuitem_get(lua_State *L)
 {
-    gui2::menu_item& item = luaW_checkmenuitem(L, 1);
-    char const *m = luaL_checkstring(L, 2);
+	gui2::menu_item& item = luaW_checkmenuitem(L, 1);
+	char const *m = luaL_checkstring(L, 2);
 
-	return_bool_optional("checkbox", item.checkbox);
+	if(dynamic_cast<gui2::multimenu_button*>(item.get_parent())) {
+		return_bool_optional("checkbox", item.checkbox);
+	} else if(dynamic_cast<gui2::menu_button*>(item.get_parent())) {
+		return_tstring_optional("details", item.details);
+		return_string_attrib("icon", item.icon);
+	} else {
+		return_bool_optional("checkbox", item.checkbox);
+		return_string_attrib("icon", item.icon);
+		return_string_optional("image", item.image);
+	}
```

Comment:
Would it be too much to ask you to rewrite all this to use the new luaW_Registry system that I created? I wouldn't count it as a blocker, but if not I'd probably rewrite it as soon as it's merged anyway, so it's probably better to do before merging.

Reply:
I've been looking at this.  I have it getting labels and tooltips.

I haven't tried yet, I'm basically just copying unit_types, but can getter handle utils::optional<bool> for example?

Also, if you look at impl_menuitem_get here you can see how I've, for example, excluded checkboxes from menu_buttons.  I'm not sure how to handle this with your new model.

Reply:
> I haven't tried yet, I'm basically just copying unit_types, but can s/getter handle `utils::optional<bool>` for example?

Yes, it can.

> Also, if you look at impl_menuitem_get here you can see how I've, for example, excluded checkboxes from menu_buttons. I'm not sure how to handle this with your new model (I guess I could just throw an error at that point, but I'd be doing the same thing over and over whereas now it just falls through to the error code).

The new model has a `*_VALID` macro which is called to determine whether the key can be accessed. It's currently used by the scenario and schedule getters, if you need an example for reference. That'll currently only affect `__dir`, so you do need to check the condition in the getter or setter as well. (Maybe it _should_ affect get/set too, actually. That seems like it would simplify a bit of code.)